src/Services/Azure/Authentication/CustomChainedCredential.cs

```diff
@@ -18,22 +20,70 @@ namespace AzureMcp.Services.Azure.Authentication;
+    public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+    {
+        _chainedCredential ??= CreateChainedCredential(tenantId);
+        return _chainedCredential.GetToken(requestContext, cancellationToken);
+    }
+
+    public override ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+    {
+        _chainedCredential ??= CreateChainedCredential(tenantId);
+        return _chainedCredential.GetTokenAsync(requestContext, cancellationToken);
+    }
+
+    private const string AuthenticationRecordEnvVarName = "AZURE_MCP_AUTHENTICATION_RECORD";
+    private const string OnlyUseBrokerCredentialEnvVarName = "AZURE_MCP_ONLY_USE_BROKER_CREDENTIAL";
+    private const string ClientIdEnvVarName = "AZURE_MCP_CLIENT_ID";
+
+    private static bool ShouldUseOnlyBrokerCredential()
+    {
+        return EnvironmentHelpers.GetEnvironmentVariableAsBool(OnlyUseBrokerCredentialEnvVarName);
+     }
+
+    private static ChainedTokenCredential CreateChainedCredential(string? tenantId)
+    {
+        string? authRecordJson = Environment.GetEnvironmentVariable(AuthenticationRecordEnvVarName);
+        AuthenticationRecord? authRecord = null;
+        if (!string.IsNullOrEmpty(authRecordJson))
+        {
+            byte[] bytes = Encoding.UTF8.GetBytes(authRecordJson);
+            using MemoryStream authRecordStream = new MemoryStream(bytes);
+            authRecord = AuthenticationRecord.Deserialize(authRecordStream);
+        }
 
-    public override ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken) =>
-        _chainedCredential.GetTokenAsync(requestContext, cancellationToken);
+        if (ShouldUseOnlyBrokerCredential())
+        {
+            return new(CreateBrowserCredential(tenantId, authRecord));
+        }
+        else
+        {
+            return new(CreateDefaultCredential(tenantId), CreateBrowserCredential(tenantId, authRecord));
+        }
+    }
 
-    private static ChainedTokenCredential CreateChainedCredential(string? tenantId) =>
-        new(CreateDefaultCredential(tenantId), CreateBrowserCredential());
+    private static string TokenCacheName = "azure-mcp-msal.cache";
 
-    private static InteractiveBrowserCredential CreateBrowserCredential() =>
-        new(new InteractiveBrowserCredentialBrokerOptions(IntPtr.Zero)
+    private static InteractiveBrowserCredential CreateBrowserCredential(string? tenantId, AuthenticationRecord? authRecord)
+    {
+        string? clientId = Environment.GetEnvironmentVariable(ClientIdEnvVarName);
+
+        IntPtr handle = WindowHandleProvider.GetWindowHandle();
+
+        return new(new InteractiveBrowserCredentialBrokerOptions(handle)
         {
-            UseDefaultBrokerAccount = true
+            UseDefaultBrokerAccount = !ShouldUseOnlyBrokerCredential() && authRecord is null,
+            ClientId = clientId,
```

Comment:
Will this set a null Client ID here if there is nothing in the env ?
There's a default client ID set in Broker Options, will a null here override that ?
@christothes

Reply:
Yes, null will override that here. So we should not set it to null if we don't intend that to flow through.

Reply:
I pushed a commit to address this issue.